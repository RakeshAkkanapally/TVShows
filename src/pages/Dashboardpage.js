import {
  Checkbox,
  CircularProgress,
  createMuiTheme,
  FormControl,
  GridList,
  GridListTile,
  GridListTileBar,
  Input,
  InputLabel,
  ListItemText,
  MenuItem,
  Select,
} from "@material-ui/core";
import React, { useEffect, useState, useContext, useRef } from "react";
import { makeStyles, Grid } from "@material-ui/core";
import showContext from "../reducers/showContext";
import Showitem from "../components/Showitem";
import ratingsData from "../data/ratings.json";
import Header from "../components/Header";
import { ThemeProvider } from "styled-components";
import { getGenres } from "../reducers/util";
import { Alert } from "@material-ui/lab";

const useStyles = makeStyles((theme) => ({
  root: {
    display: "flex",
    flexWrap: "wrap",
    justifyContent: "space-around",
    overflow: "hidden",
  },
  formControl: {
    backgroundColor: "#fff",
    margin: theme.spacing(1),
    minWidth: theme.spacing(20),
    maxWidth: theme.spacing(30),
    borderRadius: theme.spacing(1),
  },
  load: {
    display: "flex",
    justifyContent: "center",
    height: "50vh",
    alignItems: "center",
  },
  message: {
    margin: theme.spacing(9),
  },
  gridList: {
    flexWrap: "nowrap",
    transform: "translateZ(0)",
  },
  titleBar: {
    background:
      "linear-gradient(to bottom, rgba(0,0,0,0.7) 0%, " +
      "rgba(0,0,0,0.3) 70%, rgba(0,0,0,0) 100%)",
  },
  "@global": {
    ".MuiGridListTile-root": {
      width: theme.spacing(50),
    },
    ".MuiInput-root": {
      padding: "0 0 0 10px",
    },
  },
}));

const theme = createMuiTheme({
  overrides: {
    MuiButton: {
      root: {
        fontSize: "2rem",
      },
    },
  },
});

const Dashboardpage = (match) => {
  const {
    filterShows,
    getfilterShows,
    loading,
    alertShow,
    alertShowMessage,
    getAllShows,
    selectedRating,
    setSelectedRating,
    selectedGenre,
    setSelectedGenre,
    searchShows,
    setSearchKey,
  } = useContext(showContext);
  const [genres, setGenres] = useState([]);
  let initialRender = useRef(true);
  const classes = useStyles();

  useEffect(async () => {
    setGenres(await getGenres());

    // eslint-disable-next-line
  }, []);

  useEffect(async () => {
    let searchValue = match.location.search.split("=")[1];

    if (typeof searchValue !== undefined && match.location.search) {
      searchShows(searchValue);
    } else {
      await getAllShows();
      setSearchKey("");
      setSelectedRating("All");
      setSelectedGenre([]);
    }
  }, [match]);

  useEffect(async () => {
    initialRender.current || typeof searchValue === undefined
      ? (initialRender.current = false)
      : await getfilterShows(selectedRating, selectedGenre);

    // eslint-disable-next-line
  }, [selectedRating, selectedGenre]);

  const handleChange = (event) => {
    setSelectedGenre(event.target.value);
  };

  const handleChangeRating = (event) => {
    setSelectedRating(event.target.value);
  };

  const ratings = ratingsData;

  return (
    <div>
      <Header />

      <Grid container justify="center">
        <Grid item sm container justify="center">
          <FormControl variant="filled" className={classes.formControl}>
            <InputLabel id="select-filled-label">Rating</InputLabel>
            <Select
              labelId="select-filled-label"
              id="select-rating"
              value={selectedRating}
              onChange={handleChangeRating}
              autoWidth
              data-testid="select-rating"
            >
              <MenuItem value="" disabled>
                <em>Please select rating</em>
              </MenuItem>
              {ratings.map((option) => (
                <MenuItem key={option.value} value={option.value}>
                  {option.label === "All" ? option.label : "> " + option.label}
                </MenuItem>
              ))}
            </Select>
          </FormControl>
          <FormControl variant="filled" className={classes.formControl}>
            <InputLabel id="mutiple-checkbox-label" className={classes.label}>
              Genres
            </InputLabel>
            <Select
              disableUnderline={true}
              labelId="mutiple-checkbox-label"
              id="mutiple-checkbox"
              multiple
              value={selectedGenre}
              onChange={handleChange}
              input={<Input />}
              renderValue={(selected) => selected.join(", ")}
              autoWidth
              
              data-testid="select-genre"
            >
              <MenuItem value="">
                <em></em>
              </MenuItem>
              {genres.map((name) => (
                <MenuItem key={name} value={name}>
                  <Checkbox checked={selectedGenre.indexOf(name) > -1} />
                  <ListItemText primary={name} />
                </MenuItem>
              ))}
            </Select>
          </FormControl>
        </Grid>
      </Grid>

      <Grid container justify="center">
        {alertShow && (
          <Alert className={classes.message} severity="error">
            {alertShowMessage}
          </Alert>
        )}
        {loading && (
          <div className={classes.load}>
            <CircularProgress />
          </div>
        )}
      </Grid>
      <div className={classes.root}>
        <GridList className={classes.gridList}>
          <ThemeProvider theme={theme}>
            {filterShows &&
              filterShows.map((item) => (
                <GridListTile className={classes.gridListTile} key={item.id}>
                  <GridListTileBar
                    title={item.name}
                    titlePosition="top"
                    classes={{
                      root: classes.titleBar,
                    }}
                  ></GridListTileBar>

                  <Showitem
                    id={item.id}
                    name={item.name}
                    url={"show"}
                    image={item.image}
                    rating={item.rating}
                    genres={item.genres}
                    rows={[
                      {
                        name: "Rating",
                        value: item.rating.average ? item.rating.average : "NA",
                      },
                      { name: "Genre", value: item.genres },
                    ]}
                    key={item.id}
                  />
                </GridListTile>
              ))}
          </ThemeProvider>
        </GridList>
      </div>
    </div>
  );
};

export default Dashboardpage;
