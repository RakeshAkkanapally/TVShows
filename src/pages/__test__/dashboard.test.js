import React from "react";

import renderer from "react-test-renderer";
import ShowContext from "../../reducers/showContext";
import Dashboardpage from "./../Dashboardpage";
import { BrowserRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import { createMuiTheme } from "@material-ui/core";


describe(" verify dashboard", () => {
 
  it("renders correctly", () => {
    const searchShows = jest.fn();
    const setSearchKey = jest.fn();
    const getAllShows = jest.fn();
    const setSelectedRating = jest.fn();
    const setSelectedGenre = jest.fn();
    const mockUseContext = jest.fn().mockImplementation(() => ({
      selectedGenre: [],
    }));
    const theme = createMuiTheme();

    React.useContext = mockUseContext;

    const tree = renderer
      .create(
        <BrowserRouter>
          <ThemeProvider theme={theme}>
            <ShowContext.Provider
              value={{
                searchShows,
                setSearchKey,
                getAllShows,
                setSelectedRating,
                setSelectedGenre,
              }}
            >
              <Dashboardpage />
            </ShowContext.Provider>
          </ThemeProvider>
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
    //   expect(screen.getByText("Hello stranger!")).toBeInTheDocument();
  });
});
