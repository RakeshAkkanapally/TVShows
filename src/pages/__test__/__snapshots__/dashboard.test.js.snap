// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div>
    <header
      className="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary makeStyles-root-7 MuiPaper-elevation4"
    >
      <div
        className="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center"
      >
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
        >
          <a
            aria-current="page"
            className="makeStyles-logo-10 active"
            href="/"
            onClick={[Function]}
            style={Object {}}
          >
            <div
              className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M18 4l2 4h-3l-2-4h-2l2 4h-3l-2-4H8l2 4H7L5 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V4h-4z"
                />
              </svg>
              <h6
                className="MuiTypography-root MuiTypography-h6"
              >
                <strong>
                  TV SHOWS
                </strong>
              </h6>
            </div>
          </a>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
        >
          <div
            className="MuiPaper-root makeStyles-paper-8 MuiPaper-elevation1 MuiPaper-rounded"
          >
            <div
              align="center"
              className="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-justify-xs-center"
            >
              <div
                className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
              >
                <div
                  className="MuiInputBase-root makeStyles-input-12 MuiInputBase-adornedEnd"
                  onClick={[Function]}
                >
                  <input
                    className="MuiInputBase-input MuiInputBase-inputAdornedEnd"
                    data-testid="search"
                    onAnimationStart={[Function]}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    placeholder="Search TV Shows.."
                    type="text"
                  />
                  <button
                    className="MuiButtonBase-root MuiIconButton-root"
                    data-testid="searchIcon"
                    disabled={false}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onDragLeave={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    tabIndex={0}
                    type="button"
                  >
                    <span
                      className="MuiIconButton-label"
                    >
                      <svg
                        aria-hidden={true}
                        className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                        />
                      </svg>
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </header>
  </div>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-justify-xs-center"
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-justify-xs-flex-end MuiGrid-grid-sm-true"
    >
      <div
        className="MuiFormControl-root makeStyles-formControl-2"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled"
          data-shrink={false}
          id="select-filled-label"
        >
          Rating
        </label>
        <div
          className="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
          onClick={[Function]}
        >
          <div
            aria-haspopup="listbox"
            aria-labelledby="select-filled-label simple-select-filled"
            className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-filled MuiInputBase-input MuiFilledInput-input"
            id="simple-select-filled"
            onBlur={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            role="button"
            tabIndex={0}
          >
            <span
              dangerouslySetInnerHTML={
                Object {
                  "__html": "&#8203;",
                }
              }
            />
          </div>
          <input
            aria-hidden={true}
            className="MuiSelect-nativeInput"
            onAnimationStart={[Function]}
            onChange={[Function]}
            required={false}
            tabIndex={-1}
          />
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconFilled"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-sm-true"
    >
      <div
        className="MuiFormControl-root makeStyles-formControl-2"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled MuiFormLabel-filled"
          data-shrink={true}
          id="mutiple-checkbox-label"
        >
          Genres
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <div
            aria-haspopup="listbox"
            aria-labelledby="mutiple-checkbox-label mutiple-checkbox"
            className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-filled MuiInputBase-input MuiInput-input"
            id="mutiple-checkbox"
            onBlur={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            role="button"
            tabIndex={0}
          >
            <span
              dangerouslySetInnerHTML={
                Object {
                  "__html": "&#8203;",
                }
              }
            />
          </div>
          <input
            aria-hidden={true}
            className="MuiSelect-nativeInput"
            onAnimationStart={[Function]}
            onChange={[Function]}
            required={false}
            tabIndex={-1}
            value=""
          />
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconFilled"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-justify-xs-center"
  />
  <div
    className="makeStyles-root-1"
  >
    <ul
      className="MuiGridList-root makeStyles-gridList-5"
      style={
        Object {
          "margin": -2,
        }
      }
    />
  </div>
</div>
`;
